src/Model/LazyBSONDocument.php

```diff
@@ -0,0 +1,235 @@
+        }
+
+        $this->codecLibrary = $codecLibrary ?? new LazyBSONCodecLibrary();
+    }
+
+    /** @return TValue */
+    public function __get(string $property)
+    {
+        $this->readFromBson($property);
+
+        if (isset($this->unset[$property]) || ! $this->exists[$property]) {
+            trigger_error(sprintf('Undefined property: %s', $property), E_USER_WARNING);
+
+            return null;
+        }
+
+        return array_key_exists($property, $this->set) ? $this->set[$property] : $this->read[$property];
+    }
+
+    public function __isset(string $name): bool
+    {
+        // If we've looked for the value, return the cached result
+        if (isset($this->exists[$name])) {
+            return $this->exists[$name];
+        }
+
+        return $this->exists[$name] = $this->bson->has($name);
+    }
+
+    /** @param TValue $value */
+    public function __set(string $property, $value): void
+    {
+        $this->set[$property] = $value;
+        unset($this->unset[$property]);
+        $this->exists[$property] = true;
+    }
+
+    public function __unset(string $name): void
+    {
+        $this->unset[$name] = true;
+        $this->exists[$name] = false;
+        unset($this->set[$name]);
+    }
+
+    /** @return Iterator<string, TValue> */
+    public function getIterator(): CallbackIterator
+    {
+        $itemIterator = new AppendIterator();
+        // Iterate through all fields in the BSON document
+        $itemIterator->append($this->bson->getIterator());
+        // Then iterate over all fields that were set
+        $itemIterator->append(new ArrayIterator($this->set));
+
+        /** @var array<string, bool> $seen */
+        $seen = [];
+
+        return new CallbackIterator(
+            // Skip keys that were unset or handled in a previous iterator
+            new CallbackFilterIterator(
+                $itemIterator,
+                /** @param TValue $current */
+                function ($current, string $key) use (&$seen): bool {
+                    return ! isset($this->unset[$key]) && ! isset($seen[$key]);
+                },
+            ),
+            /**
+             * @param TValue $value
+             * @return TValue
+             */
+            function ($value, string $key) use (&$seen) {
+                // Mark key as seen, skipping any future occurrences
+                $seen[$key] = true;
+
+                // Return actual value (potentially overridden by __set)
+                return $this->__get($key);
+            },
+        );
+    }
+
+    /** @param mixed $offset */
+    public function offsetExists($offset): bool
+    {
+        return $this->__isset((string) $offset);
+    }
+
+    /**
+     * @param mixed $offset
+     * @return TValue
+     */
+    #[ReturnTypeWillChange]
+    public function offsetGet($offset)
+    {
+        return $this->__get((string) $offset);
```

Comment:
You may trigger a warning if `$offset` is an `int`, they are not valid field names. Or there is an exception later in the call to `__get`?

Reply:
There is no exception - we always cast to string. `LazyBSONArray` only accepts numeric arguments and triggers a warning about undefined offsets. I've made this consistent and any non-string offsets will now be considered invalid and rejected.